File: pipeline/src/sterish_pipeline/content_hash.py
"""Sterish ``content_hash`` v1 — the frozen canonical algorithm.

Normative spec: ``docs/specs/content-hash.md`` (FROZEN, STE-10). Reference
implementation: ``docs/specs/reference/content_hash.py``. If this file and the
spec ever disagree, the spec wins and this file is the bug.

The hash is the byte identity of a skill. It is computed in three places
(pipeline / Rust contract / TypeScript client) and they must agree byte-for-byte,
or ``check(skill)`` lies. This module is the pipeline's implementation; it is
verified against the same shared vectors as the Rust and TypeScript ones.

    CANON = MAGIC
         || u32be(file_count)
         || for each file, sorted ASC bytewise by path_bytes:
                u32be(len(path_bytes))   || path_bytes
                u32be(len(norm_content)) || norm_content

    MAGIC        = b"sterish-content-hash/v1\\n"     (24 bytes, includes the \\n)
    content_hash = sha256(CANON)                    (32 bytes, 64 lowercase hex)

``norm_content`` normalizes line endings (CRLF -> LF, CR -> LF) and strips
trailing newlines; files must be valid UTF-8. skill_id and version are NOT part
of the hash — only the file paths and their normalized bytes are.
"""

from __future__ import annotations

import hashlib
import struct
from collections.abc import Mapping, Sequence
from pathlib import Path

#: Domain-separation prefix. 24 bytes, trailing newline included.
MAGIC = b"sterish-content-hash/v1\n"
assert len(MAGIC) == 24, "MAGIC must be exactly 24 bytes"

#: Dropped by the packager BEFORE hashing (not part of the algorithm).
EXCLUDED_DIRS = frozenset({".git", "node_modules", "__pycache__", ".venv", "target"})
EXCLUDED_FILES = frozenset({".DS_Store"})
EXCLUDED_SUFFIXES = (".pyc",)


class ContentHashError(ValueError):
    """Base class. ``kind`` is the stable, cross-language error name."""

    kind = "ContentHashError"


class EmptyFileSet(ContentHashError):
    kind = "EmptyFileSet"


class DuplicatePath(ContentHashError):
    kind = "DuplicatePath"


class InvalidPath(ContentHashError):
    kind = "InvalidPath"


class NotUtf8(ContentHashError):
    kind = "NotUtf8"


def _u32be(n: int) -> bytes:
    if n < 0 or n > 0xFFFF_FFFF:
        raise ContentHashError(f"value out of u32 range: {n}")
    return struct.pack(">I", n)


def check_path(path_bytes: bytes) -> None:
    """Reject anything that is not a clean, relative, POSIX path."""
    if not path_bytes:
        raise InvalidPath("empty path")
    try:
        text = path_bytes.decode("utf-8")
    except UnicodeDecodeError as exc:
        raise InvalidPath("path is not valid UTF-8") from exc
    if "\\" in text:
        # A backslash is a legal POSIX filename byte, but rejected on purpose so
        # a Windows-style path never becomes one filename and hashes differently
        # than the same skill packaged on another OS.
        raise InvalidPath(f"backslash is not a path separator: {text!r}")
    if "\x00" in text:
        raise InvalidPath(f"NUL byte in path: {text!r}")
    for part in text.split("/"):
        if part == "":
            raise InvalidPath(f"empty path component (leading/trailing/double slash): {text!r}")
        if part == ".":
            raise InvalidPath(f"'.' component not allowed: {text!r}")
        if part == "..":
            raise InvalidPath(f"'..' component not allowed: {text!r}")


def normalize_content(raw: bytes) -> bytes:
    """v1 content normalization: CRLF -> LF, CR -> LF, strip trailing LF."""
    try:
        raw.decode("utf-8")
    except UnicodeDecodeError as exc:
        raise NotUtf8(f"content is not valid UTF-8: {exc}") from exc
    normalized = raw.replace(b"\r\n", b"\n").replace(b"\r", b"\n")
    return normalized.rstrip(b"\n")


def hash_bytes(data: bytes) -> str:
    """SHA-256 of raw bytes as 64 lowercase hex characters (per-file digest)."""
    return hashlib.sha256(data).hexdigest()


def _normalize_files(files: Mapping[str, bytes] | Sequence[tuple[str, bytes]]):
    pairs = files.items() if isinstance(files, Mapping) else files
    seen: set[bytes] = set()
    items: list[tuple[bytes, bytes]] = []
    for path, raw in pairs:
        path_bytes = path.encode("utf-8") if isinstance(path, str) else path
        check_path(path_bytes)
        if path_bytes in seen:
            raise DuplicatePath(f"duplicate path: {path_bytes!r}")
        seen.add(path_bytes)
        items.append((path_bytes, normalize_content(raw)))
    if not items:
        raise EmptyFileSet("a skill must contain at least one file")
    # ASC bytewise on the RAW path bytes (Python bytes ordering is bytewise).
    items.sort(key=lambda item: item[0])
    return items


def canonical_bytes(files: Mapping[str, bytes] | Sequence[tuple[str, bytes]]) -> bytes:
    """Build CANON from ``{path: raw_bytes}``. Input order is irrelevant."""
    items = _normalize_files(files)
    buf = bytearray(MAGIC)
    buf += _u32be(len(items))
    for path_bytes, content in items:
        buf += _u32be(len(path_bytes))
        buf += path_bytes
        buf += _u32be(len(content))
        buf += content
    return bytes(buf)


def content_hash(files: Mapping[str, bytes] | Sequence[tuple[str, bytes]]) -> str:
    """Compute ``content_hash`` for an in-memory skill. 64 lowercase hex chars.

    Args:
        files: ``{relative_path: raw_file_bytes}``, at least one entry. skill_id
            and version are intentionally NOT arguments — they are not part of
            the byte identity of the skill (see the frozen spec).
    """
    return hash_bytes(canonical_bytes(files))


def _is_excluded(rel_parts: Sequence[str], name: str) -> bool:
    if any(part in EXCLUDED_DIRS for part in rel_parts):
        return True
    if name in EXCLUDED_FILES:
        return True
    return name.endswith(EXCLUDED_SUFFIXES)


def read_skill_files(root: Path | str) -> dict[str, bytes]:
    """Read every hashable file under ``root`` into ``{path: raw_bytes}``.

    Applies the packager exclusion list. Symlinks are skipped: they carry no
    content of their own and would let a skill hash bytes outside its root.
    """
    root_path = Path(root)
    if root_path.is_file():
        return {root_path.name: root_path.read_bytes()}
    if not root_path.is_dir():
        raise ContentHashError(f"not a file or directory: {root_path}")

    files: dict[str, bytes] = {}
    for path in sorted(root_path.rglob("*")):
        if path.is_symlink() or not path.is_file():
            continue
        rel = path.relative_to(root_path)
        rel_parts = list(rel.parts[:-1])
        if _is_excluded(rel_parts, path.name):
            continue
        files[rel.as_posix()] = path.read_bytes()

    if not files:
        raise EmptyFileSet(f"no hashable files found under {root_path}")
    return files


def content_hash_path(root: Path | str) -> str:
    """Compute ``content_hash`` for a skill on disk."""
    return content_hash(read_skill_files(root))

File: frontend/src/lib/contentHash.ts
/**
 * `content_hash` v1 in the browser, so a user can check bytes before installing.
 *
 * A port of `docs/specs/reference/contentHash.ts`, the frozen reference that
 * `scripts/verify-content-hash.sh` holds identical to the Python and Rust
 * implementations. The algorithm is copied, not reinterpreted: if this file and
 * `docs/specs/content-hash.md` ever disagree, the spec wins and this file is the
 * bug. `contentHash.test.ts` runs the shared vectors against it to keep it that
 * way.
 *
 * Two differences from the reference, both forced by the browser:
 *
 * - sha256 comes from WebCrypto, which is async, so `contentHash` returns a
 *   promise. `node:crypto` does not exist here.
 * - There is no filesystem walk. The caller hands over files picked or dropped
 *   by the user, and `packageFiles` applies the packager's exclusion list
 *   (spec section 1.3) to them instead.
 */

const ENCODER = new TextEncoder();
const STRICT_UTF8 = new TextDecoder("utf-8", { fatal: true });

/** Domain separation prefix. 24 bytes, trailing newline included. */
export const MAGIC: Uint8Array = ENCODER.encode("sterish-content-hash/v1\n");

export type ContentHashErrorKind =
  | "EmptyFileSet"
  | "DuplicatePath"
  | "InvalidPath"
  | "NotUtf8";

export class ContentHashError extends Error {
  readonly kind: ContentHashErrorKind;

  constructor(kind: ContentHashErrorKind, message: string) {
    super(`${kind}: ${message}`);
    this.name = "ContentHashError";
    this.kind = kind;
  }
}

export interface SkillFile {
  /** Skill-root-relative POSIX path, as text. */
  path: string;
  /** Raw file bytes, before normalisation. */
  raw: Uint8Array;
}

function u32be(n: number): Uint8Array {
  if (!Number.isInteger(n) || n < 0 || n > 0xffff_ffff) {
    throw new Error(`value out of u32 range: ${n}`);
  }
  return new Uint8Array([
    (n >>> 24) & 0xff,
    (n >>> 16) & 0xff,
    (n >>> 8) & 0xff,
    n & 0xff,
  ]);
}

/**
 * Bytewise order on the raw UTF-8 path. Not `localeCompare` and not the default
 * string sort, which is UTF-16 code unit order and disagrees for non-BMP code
 * points (the `non-bmp-path-order` vector).
 */
function compareBytes(a: Uint8Array, b: Uint8Array): number {
  const n = Math.min(a.length, b.length);
  for (let i = 0; i < n; i++) {
    if (a[i] !== b[i]) return a[i] < b[i] ? -1 : 1;
  }
  return a.length - b.length;
}

function isUtf8(bytes: Uint8Array): boolean {
  try {
    STRICT_UTF8.decode(bytes);
    return true;
  } catch {
    return false;
  }
}

function checkPath(path: string): void {
  if (path.length === 0) {
    throw new ContentHashError("InvalidPath", "empty path");
  }
  if (path.includes("\\")) {
    throw new ContentHashError(
      "InvalidPath",
      `backslash is not a path separator: ${path}`,
    );
  }
  if (path.includes("\0")) {
    throw new ContentHashError("InvalidPath", `NUL byte in path: ${path}`);
  }
  for (const part of path.split("/")) {
    if (part === "") {
      throw new ContentHashError(
        "InvalidPath",
        `empty path component (leading, trailing or double slash): ${path}`,
      );
    }
    if (part === "." || part === "..") {
      throw new ContentHashError(
        "InvalidPath",
        `'${part}' component not allowed: ${path}`,
      );
    }
  }
}

/**
 * (a) every CRLF to LF, (b) every remaining CR to LF, (c) strip all trailing
 * LF. Three literal passes so it maps line for line onto the spec text.
 */
export function normalizeContent(raw: Uint8Array): Uint8Array {
  if (!isUtf8(raw)) {
    throw new ContentHashError("NotUtf8", "content is not valid UTF-8");
  }
  const CR = 0x0d;
  const LF = 0x0a;

  const out = new Uint8Array(raw.length);
  let n = 0;
  for (let i = 0; i < raw.length; ) {
    if (raw[i] === CR && i + 1 < raw.length && raw[i + 1] === LF) {
      out[n++] = LF;
      i += 2;
    } else {
      out[n++] = raw[i];
      i += 1;
    }
  }
  for (let i = 0; i < n; i++) {
    if (out[i] === CR) out[i] = LF;
  }
  let end = n;
  while (end > 0 && out[end - 1] === LF) end--;
  return out.subarray(0, end);
}

export function canonicalBytes(files: SkillFile[]): Uint8Array {
  if (files.length === 0) {
    throw new ContentHashError(
      "EmptyFileSet",
      "a skill must contain at least one file",
    );
  }

  const seen = new Set<string>();
  const items = files.map((file) => {
    checkPath(file.path);
    if (seen.has(file.path)) {
      throw new ContentHashError(
        "DuplicatePath",
        `duplicate path: ${file.path}`,
      );
    }
    seen.add(file.path);
    return {
      pathBytes: ENCODER.encode(file.path),
      content: normalizeContent(file.raw),
    };
  });

  items.sort((a, b) => compareBytes(a.pathBytes, b.pathBytes));

  const chunks: Uint8Array[] = [MAGIC, u32be(items.length)];
  for (const item of items) {
    chunks.push(u32be(item.pathBytes.length), item.pathBytes);
    chunks.push(u32be(item.content.length), item.content);
  }
  const buf = new Uint8Array(chunks.reduce((sum, c) => sum + c.length, 0));
  let offset = 0;
  for (const chunk of chunks) {
    buf.set(chunk, offset);
    offset += chunk.length;
  }
  return buf;
}

/** 64 lowercase hex characters, exactly what `GET /check/by-hash` expects. */
export async function contentHash(files: SkillFile[]): Promise<string> {
  const digest = await crypto.subtle.digest(
    "SHA-256",
    canonicalBytes(files) as Uint8Array<ArrayBuffer>,
  );
  return Array.from(new Uint8Array(digest), (b) =>
    b.toString(16).padStart(2, "0"),
  ).join("");
}

const EXCLUDED_DIRS = new Set([
  ".git",
  "node_modules",
  "__pycache__",
  ".venv",
  "target",
]);
const EXCLUDED_FILES = new Set([".DS_Store"]);
const EXCLUDED_SUFFIXES = [".pyc"];

/**
 * Spec section 1.3: files the packager drops before hashing. They are not part
 * of the algorithm, so an uploaded folder that still contains `.git/` must have
 * it removed here or it would hash to bytes nobody ever audited.
 */
export function isExcludedPath(path: string): boolean {
  const parts = path.split("/");
  const name = parts[parts.length - 1];
  return (
    parts.slice(0, -1).some((dir) => EXCLUDED_DIRS.has(dir)) ||
    EXCLUDED_FILES.has(name) ||
    EXCLUDED_SUFFIXES.some((suffix) => name.endsWith(suffix))
  );
}

/**
 * Turn what a browser picker produced into skill-root-relative paths.
 *
 * A folder picked with `webkitdirectory` reports `webkitRelativePath` as
 * `<folder>/<path>`, and the folder name is not part of the skill: the
 * reference hashes paths relative to the skill root, so the first component is
 * removed. Loose files have no relative path and use their own name.
 */
export function relativeSkillPath(file: {
  name: string;
  webkitRelativePath?: string;
}): string {
  const relative = file.webkitRelativePath;
  if (!relative) return file.name;
  const slash = relative.indexOf("/");
  return slash === -1 ? relative : relative.slice(slash + 1);
}
